presto-main/src/main/java/com/facebook/presto/operator/HashBuilderOperator.java

```diff
@@ -366,6 +376,13 @@ private void spillInput(Page page)
         checkState(spillInProgress.isDone(), "Previous spill still in progress");
         checkSpillSucceeded(spillInProgress);
         spillInProgress = getSpiller().spill(page);
+
+        // check that spilled data can still fit into memory limit as otherwise
+        // it fails later during unspilling when all spilled pages need to loaded into memory
+        if (getSpiller().getSpilledPagesInMemorySize() > maxUserMemoryBytes) {
```

Comment:
You can also get `query.max-memory-per-node` value from the session directly using `getQueryMaxMemoryPerNode(operatorContext.getSession()).toBytes()`. With this, you wont have to pass this value around.

Reply:
I made it this way intentionally as `QueryContext` seems to be the only source of truth for max memory. In case `resource_overcommit=true` the max memory may be more than `query.max-memory-per-node` (see how memory limits are set https://github.com/prestodb/presto/blob/ed2a648b46225860260180aad47a0a7e853f366f/presto-main/src/main/java/com/facebook/presto/execution/SqlTaskManager.java#L411). Pulling the limit directly from the session would disallow this overcommit use-case.

Reply:
Actually, we do not use SqlTaskManager in Sapphire query execution flow so this call to resourceOvercommit will not be invoked in Sapphire at all. Sapphire uses Spark backend for task scheduling and management. If you see, there is a class called `PrestoSparkTaskManager` which extends TaskManager but provides no implementation for any methods. This was probably added to probably make unit tests work in Sapphire.

Reply:
> Actually, we do not use SqlTaskManager in Sapphire query execution flow so this call to resourceOvercommit will not be invoked in Sapphire at all.

I see, makes sense. But we're making change here in the Presto codebase in general. I'd argue the fix should be compatible with any use case.

Reply:
+1 with what @pgupta2 has mentioned. Plus, `resource_overcommit` is really something we probably would avoid to use.

Reply:
OK, I'll simplify the code in the next revision.